---
title: "RNA-seq analysis workflow"
output: html_document
date: "2024-08-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Load in the required libraries, if you do not have them you need to install them
```{r}
library(tximportData)
library(rhdf5)
library(readr)
library(tximport)
library(DESeq2)
library(tidyverse)
library(ggvenn)
library("UpSetR")
library("ComplexHeatmap")
library(clusterProfiler)
library("org.Ce.eg.db")
```

#Get the inputs and set output directory, counts should be in tsv format with '_abundance.tsv'
#Simple change these paths to the ones on your laptop
```{r}
counts_directory_path <- "/Users/steve/Ollie-rnaSeq/kallisto-counts"
count_directory_name <- "kallisto-counts/" # (This is need to split the file path string)
sample_sheet_path <- "/Users/steve/Ollie-rnaSeq/kallisto-counts/sample_sheet_spr1-5.csv"
output_dir <- "/Users/steve/Ollie-rnaSeq/DEG/"
```

```{r}
#Gather the files
files <- list.files(counts_directory_path, pattern="*_abundance.tsv", full.names=TRUE)

#Just get the file names without abunance.tsv
filenames <- str_split_fixed(files, count_directory_name, 2)
filenames2 <- str_split_fixed(filenames[,2], '_abundance.', 2)
#Name the files by the filenames
names(files) <- filenames2[,1]
#Get a table of gene transcripts against their gene
gene_transcripts <- read.table(file.path("/Users/steve/Index-files/annotation_files/c.elegans.PRJNA13758.WS285_cannonical","c_elegans.PRJNA13758.WS285.canonical_geneset.gene_transcript_id.sorted.txt"))
#Flip the columns of the transcripts table
txdb <- gene_transcripts[,c(2,1)]
#Create the txi object to hold the counts table
txi.kallisto <- tximport(files, type = "kallisto", tx2gene = txdb) #Tx to gene
#Read in the sample sheet ***N.B sample must be in the same order as the txi object, it will not give an error when
#carrying out the differential analysis if they are not!!***
samples <- read.table(file.path(sample_sheet_path), header=TRUE,sep=',')
```

#Carry out DESeq2 DE with feature kallisto data
```{r}
#Create DESeq dataset from txi object
dds_txi <- DESeqDataSetFromTximport(txi.kallisto,
                                   colData = samples,
                                   design = ~ Gene)

#Carry out the differential analysis
dds_txi <- DESeq(dds_txi)
```


#Carry out variance stabilisation to correction for size factors or normalization factors (Needed for PCA)
#This uses the vsd function that uses a parametric fit and the rlog function (rlog used to make heatmaps later)
```{r}
vsd_txi <- vst(dds_txi, blind=FALSE)
rlog_dds_txi <- rlog(dds_txi, blind=FALSE)
mat_df_all_txi <-counts(dds_txi, normalized=TRUE)
```


#PCA plot to check the clustering of samples
#They should show some clustering by sample type
```{r}
pcaData <- plotPCA(rlog_dds_txi, intgroup=c("Gene", "Sample"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Gene, label=Sample)) +
   geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed()
```


#Get a dataframe of normalised counts, this can be used to check normalisation or make plots of the 
#normalised counts
```{r}
normalized_counts <- counts(dds_txi, normalized=T) %>% 
                     data.frame() %>%
                     rownames_to_column(var="gene") 
```


#Set the p-value cutoff for DEG lists
```{r}
pvaluecutoff <- 0.001
```

#Contrast the experimental conditions, this code block can be copied multiple time to contrast different conditions. Make sure you change the all the object names from 'contrast1'
```{r}
contrast1 <- results(dds_txi,
                     contrast=c("Gene", "hpl2", "WT"))

#Get the genes that pass the p-value threshold and order them descending
contrast1_genes <- subset(contrast1, padj< pvaluecutoff & !is.na(padj))
contrast1_genes <- contrast1_genes[order(contrast1_genes$log2FoldChange, decreasing = TRUE),]

#Get the DE genes that are up and down-regulated, the down-regulated genes are re-ordered so they
#are ascending
contrast1_genes_up <- subset(contrast1_genes, log2FoldChange > 0 )
contrast1_genes_down <- subset(contrast1_genes, log2FoldChange < 0 )
contrast1_genes_down <- contrast1_genes_down[order(contrast1_genes_down$log2FoldChange, decreasing = FALSE),]


write.csv(contrast1_genes,file.path(output_dir,"/contrast1_DEG.csv"),quote=FALSE)
write.csv(contrast1_genes_up,file.path(output_dir,"/contrast1_UP-DEG.csv"),quote=FALSE)
write.csv(contrast1_genes_down,file.path(output_dir,"/contrast1_UP-DEG.csv"),quote=FALSE)print("Number of differential genes")
nrow(contrast1_genes)
```

```{r}
contrast2 <- results(dds_txi,
                     contrast=c("Gene", "spr1", "WT"))

#Get the genes that pass the p-value threshold and order them descending
contrast2_genes <- subset(contrast1, padj< pvaluecutoff & !is.na(padj))
contrast2_genes <- contrast1_genes[order(contrast2_genes$log2FoldChange, decreasing = TRUE),]

#Get the DE genes that are up and down-regulated, the down-regulated genes are re-ordered so they
#are ascending
contrast2_genes_up <- subset(contrast2_genes, log2FoldChange > 0 )
contrast2_genes_down <- subset(contrast2_genes, log2FoldChange < 0 )
contrast2_genes_down <- contrast2_genes_down[order(contrast2_genes_down$log2FoldChange, decreasing = FALSE),]


write.csv(contrast2_genes,file.path(output_dir,"/contrast2_DEG.csv"),quote=FALSE)
write.csv(contrast2_genes_up,file.path(output_dir,"/contrast2_UP-DEG.csv"),quote=FALSE)
write.csv(contrast2_genes_down,file.path(output_dir,"/contrast2_UP-DEG.csv"),quote=FALSE)

print("Number of differential genes")
nrow(contrast2_genes)
```

#The code below makes a MA and a volcano plot of any of the DEG objects from contrasted conditions above
```{r}
DEG_dds_object <- contrast1 # dds object from contrast above
count_columns <- c(9,10,1,2) # Columns from rlog_dds_mf for samples you want in the contrast to plot

#Get a MA plot of DEG
plotMA(DEG_dds_object, ylim=c(-10,10))

#Get a dataframe from object for this contrast an only keep genes that pass the p-value test
DEG_df <- as.data.frame(DEG_dds_object)
DEG_df_pvalue <- subset(DEG_df, padj< pvaluecutoff & !is.na(padj))

DEG_dds_tibble <- DEG_dds_object %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

DEG_dds_tibble <- DEG_dds_tibble %>% 
                  dplyr::mutate(threshold_OE = padj < 0.1 & abs(log2FoldChange) >= 0.58)

ggplot(DEG_dds_tibble) +
    geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold_OE)) +
    ggtitle("fL1 WT vs hpl2 ") +
    xlab("log2 fold change") + 
    ylab("-log10 adjusted p-value") +
    scale_y_continuous(limits = c(0,50)) +
    theme(legend.position = "none",
          plot.title = element_text(size = rel(1.5), hjust = 0.5),
          axis.title = element_text(size = rel(1.25)))
```

#Venn diagrams of contrasts, add n the contrasts from above
```{r}
sL1_list <- list(
  WTvshpl2 = rownames(contrast1_genes), 
  WTvsspr1 = rownames(contrast2_genes))
ggvenn(
  sL1_list, 
  fill_color = c("#0073C2FF", "#EFC000FF"),
  stroke_size = 0.5, set_name_size = 4
  )
```

#Upset plots for all DEG, plus up and down DEG
#Simply add in your contrasts from above
```{r}
upset_list <- list(  WTvshpl2 = rownames(contrast1_genes), 
                  WTvsspr1 = rownames(contrast2_genes))

up_upset_list <- list(  WTvshpl2 = rownames(contrast1_genes_up), 
                  WTvsspr1 = rownames(contrast2_genes_up))

down_upset_list <- list(  WTvshpl2 = rownames(contrast1_genes_down), 
                  WTvsspr1 = rownames(contrast2_genes_down))

#List to matrix and make combination matrix functions taken from complexheatmap package
#upset_matrix <- list_to_matrix(upset_list)
comb_mat = make_comb_mat(upset_list)
comb_mat_up = make_comb_mat(up_upset_list)
comb_mat_down = make_comb_mat(down_upset_list)

UpSet(comb_mat)
UpSet(comb_mat_up)
UpSet(comb_mat_down)
```

#Carry out enrichment analysis
#This analysis uses clusterprofiler, wormcat may be a better option
#WormCat is available here: www.wormcat.com
#N.B Anna Protasio is working on a WormCat method in R
```{r}
DEG_dds_oject <- contrast1 # contrast from contrast above, simply change the input here

#Get a dataframe from object for this contrast
DEG_df <- as.data.frame(DEG_dds_oject)

#Get the best DEG
DEG_df_best <- subset(DEG_df, padj< pvaluecutoff & !is.na(padj))

#Order the best DEG from top to bottom
DEG_df_best <- DEG_df_best[order(DEG_df_best$log2FoldChange, decreasing = TRUE),]

DEG_df_up <-  DEG_df_best[DEG_df_best$log2FoldChange > 0,]
DEG_df_down <-  DEG_df_best[DEG_df_best$log2FoldChange < 0,]
DEG_df_down <- DEG_df_bottom[order(DEG_df_bottom$log2FoldChange, decreasing = FALSE),]

genes_all <- rownames(DEG_df_best)
genes_down <- rownames(DEG_df_down)
genes_top <- rownames(DEG_df_top)

#GO over-representation analysis, change the input to either genes_all, genes_down or genes_up
ego <- enrichGO(gene          = genes_all,
                OrgDb         = org.Ce.eg.db,
                keyType       = 'WORMBASE',
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = TRUE)
goplot(ego)
```

#Compare two gene lists, give differnt contrasts as inputs to compare different conditions
```{r}
#Get a list of DEG from the first contrast
DEG_list1 <- contrast1
DEG_df1 <- as.data.frame(DEG_list1)
DEG_df1 <- subset(DEG_df1, padj< pvaluecutoff & !is.na(padj))

#Get another list of DEG from another contrast
DEG_list2 <- contrast2
DEG_df2 <- as.data.frame(DEG_list2)
DEG_df2 <- subset(DEG_df2, padj< pvaluecutoff & !is.na(padj))

#Find the genes that are present in both lists
DEG_inboth <- subset(DEG_df1, rownames(DEG_df1) %in% rownames(DEG_df2))
genes_inboth <- rownames(DEG_inboth)
print("Number of genes in both lists:")
length(genes_inboth)
```

